lms/djangoapps/course_home_api/outline/views.py

```diff
@@ -375,6 +381,188 @@ def finalize_response(self, request, response, *args, **kwargs):
+            is_masquerading=user_is_masquerading,
+        )
+        if courseware_mfe_navigation_sidebar_blocks_caching_is_enabled():
+            course_blocks = cache.get(cache_key)
+            cached = course_blocks is not None
+        else:
+            cached = False
+            course_blocks = None
+
+        if not course_blocks:
+            if getattr(enrollment, 'is_active', False) or bool(staff_access):
+                course_blocks = get_course_outline_block_tree(request, course_key_string, request.user)
+            elif allow_public_outline or allow_public or user_is_masquerading:
+                course_blocks = get_course_outline_block_tree(request, course_key_string, None)
+
+            if courseware_mfe_navigation_sidebar_blocks_caching_is_enabled():
+                cache.set(cache_key, course_blocks, self.COURSE_BLOCKS_CACHE_TIMEOUT)
+
+        course_blocks = self.filter_unavailable_blocks(course_blocks, course_key)
+
+        if cached:
+            # If the data was cached, we need to mark the blocks as complete or not complete.
+            course_blocks = self.mark_complete_recursive(course_blocks)
+
+        context = self.get_serializer_context()
+        context.update({
+            'include_vertical': True,
+            'extra_fields': ['special_exam_info'],
+            'enable_prerequisite_block_type': True,
+        })
+
+        serializer = self.get_serializer_class()(course_blocks, context=context)
+
+        return Response(serializer.data)
+
+    def filter_unavailable_blocks(self, course_blocks, course_key):
+        """
+        Filter out sections and subsections that are not available to the current user.
+        """
+        if course_blocks:
+            user_course_outline = get_user_course_outline(course_key, self.request.user, datetime.now(tz=timezone.utc))
+            course_sections = course_blocks.get('children', [])
+            course_blocks['children'] = self.get_available_sections(user_course_outline, course_sections)
+
+            for section_data in course_sections:
+                section_data['children'] = self.get_available_sequences(
+                    user_course_outline,
+                    section_data.get('children', [])
+                )
+                accessible_sequence_ids = {str(usage_key) for usage_key in user_course_outline.accessible_sequences}
+                for sequence_data in section_data['children']:
+                    sequence_data['accessible'] = sequence_data['id'] in accessible_sequence_ids
+
+        return course_blocks
+
+    def mark_complete_recursive(self, block):
+        """
+        Mark blocks as complete or not complete based on the completions_dict.
+        """
+        if 'children' in block:
+            block['children'] = [self.mark_complete_recursive(child) for child in block['children']]
+            block['complete'] = all(child['complete'] for child in block['children'] if child['type'] != 'discussion')
```

Comment:
No, because in our case, block is not an XBlock object, but a dict with block data, and therefore we cannot get the value of the `completion_mode` field here.

Reply:
Sorry, let me ask this another way: Why is the "discussion" block explicitly excluded here?

Reply:
The "discussion" block is explicitly excluded here because it cannot be marked as completed, in which case units that have a "discussion" block will never be marked as completed either.

I also considered the option of checking the complete status for blocks that have `has_score=True`, but in this case it would only apply to problems, which is also not entirely correct.

Reply:
Wouldn't this get the right tags then?

```python
from xblock.core import XBlock
from xblock.completable import XBlockCompletionMode

completable_tags = {
    tag for (tag, cls) in XBlock.load_classes()
    if XBlockCompletionMode.get_mode(cls) == XBlockCompletionMode.COMPLETABLE
}
```

Reply:
Oh, thank you very much, I didn't know that we could get completable block types in this way. 
Added this change.